subt/simulation.py

```diff
@@ -0,0 +1,159 @@
+        self.names.append(name)
+        return len(self.names)-1
+
+    @color.dodger_blue_1
+    def write(self, stream_id, data, dt=None):
+        #log(self.names[stream_id], data)
+        dt = datetime.datetime.now(datetime.timezone.utc) - self.start_time
+        return dt
+
+
+class Simulation:
+
+    def __init__(self, bus, end_condition):
+        bus.register('scan', 'rot', 'orientation', 'sim_time_sec', 'origin', 'pose2d', 'acc', 'artf')
+        self.bus = bus
+        self.end_condition = end_condition
+        self._handlers = collections.defaultdict(lambda: self.on_default)
+        self.set_handler('request_origin', self.on_request_origin)
+        self.set_handler('desired_speed', self.on_desired_speed)
+        self.set_handler('pose2d', self.on_pose)
+        self.set_handler('pose3d', self.on_pose)
+        self.origin = [7.0, 3.0, 0.0]
+        self.xyz = [0.0, 0.0, 0.0]
+        self.orientation = quaternion.identity()
+        self.speed_forward = 0.0
+        self.speed_angular = 0.0
+        self.time = datetime.timedelta()
+
+    def start(self):
+        self.thread = Thread(target=self.main)
+        self.thread.start()
+
+    def join(self, timeout=None):
+        self.thread.join(timeout)
+
+    @color.green
+    def main(self):
+        log("started")
+        self.on_request_origin(datetime.timedelta(), "request_origin", True)
+
+        for _ in range(500):
+            packet = self.bus.listen()
+            self.handle_incomming(*packet)
+            if self.end_condition(self):
+                return
+
+    def set_handler(self, channel, handler):
+        self._handlers[channel] = handler
+
+    def handle_incomming(self, dt, channel, data):
+        self._handlers[channel](dt, channel, data)
+
+    @color.red
+    def on_default(self, dt, channel, data):
+        log(dt, channel, data)
+
+    @color.gold_1
+    def on_request_origin(self, dt, channel, data):
+        log(dt, channel, data)
+        # start sending origin data until robot arrives within 0.3m from (0,0,0)
+        corrected = [rr - oo for rr, oo in zip(self.xyz, self.origin)]
+        self.bus.publish('origin', [b'X0F100L', *corrected, *self.orientation])
+        self.bus.publish('scan', [0] * 100)
+
+    @color.turquoise_4
+    def on_pose(self, dt, channel, data):
+        if channel == 'pose2d':
+            out = str(data)
+        elif channel == 'pose3d':
+            out0 = "[" + ', '.join([f"{p:.2f}" for p in data[0]]) + "]"
+            out1 = " [" + ', '.join([f"{p:.2f}" for p in data[1]]) + "]"
+            out = out0 + out1
+        log(dt, channel, out)
+
+    @color.sea_green_3
+    def on_desired_speed(self, dt, channel, data):
+        forward = data[0]/1000.
+        angular = data[1]/100.
+        diff_forward = clip(-0.1, forward - self.speed_forward, 0.1)
+        diff_angular = clip(-10.0, angular - self.speed_angular, 10.0)
+        #print(diff_forward, diff_angular)
+        self.speed_forward += diff_forward
```

Comment:
diff_forward seems to basically be acceleration, but the unit is m/s/function_call? What do you think about somehow turning it into m/s/s? Similar for the rotation.